recipe/2024-01-support-private-terraform-repository.md

```diff
@@ -0,0 +1,318 @@
+      }
+    }
+  }
+}
+
+```
+E.g of secretstore using the existing kubernetes secret.
+```diff
+resource existingSecret 'Applications.Core/secretStores@2023-10-01-preview' = {
+  name: 'existing-secret'
+  properties:{
+-  	app: app.id
+    resource: '<namespace>/<existing-secret-name>' // Reference to the name of a an external secret store
+    type: 'generic' // The type of secret in your resource
+    data: {
+      // The keys in this object are the names of the secrets in an external secret store
+      'tls.pat': {}
+      'tls.username': {}
+    }
+  }
+}
+```
+SecretStore resource also provides an option to use the existing secret, which makes it better way store credentials. But todays secret store implementation is tied to application scope and in this case, secretStore needs to be created before application and environment creation. So we need to change scope of secret store resource to global (by removing the required flag for application property).
+And the referenced secret is retrieved from the secretStore in terraform recipe driver and secret details i.e personal-access-token and username as used to modify the template path to the below format before writing it to the terraform config:
+```
+"git::https://{username}:{PERSONAL_ACCESS_TOKEN}@example.com.com/test-private-repo.git"
+```
+that will be used for recipe deployment.
+
+### API design (if applicable)
+***Model changes***
+
+Addition of new property `recipeConfig` to environment properties.
+
+```diff
+@doc("Environment properties")
+model EnvironmentProperties {
+  @doc("The status of the asynchronous operation.")
+  @visibility("read")
+  provisioningState?: ProvisioningState;
+
+  @doc("The compute resource used by application environment.")
+  compute: EnvironmentCompute;
+
+  @doc("Cloud providers configuration for the environment.")
+  providers?: Providers;
+
+  @doc("Simulated environment.")
+  simulated?: boolean;
+
+  @doc("Specifies Recipes linked to the Environment.")
+  recipes?: RecipeProperties;
+
++  @doc("Specifies recipe configurations needed for the recipes.")
++  recipeConfig?: RecipeConfigProperties;
+
+  @doc("The environment extension.")
+  @extension("x-ms-identifiers", [])
+  extensions?: Array<Extension>;
+}
+
++model RecipeConfigProperties {
++  @doc(Specifies the terraform config properties)
++  terraform?: Record<TerraformConfig>;
++}
+
++model TerraformConfig{
++  @doc(Specifies git credentials information for terraform module repository.)  
++  gitCredentials?: Record<Secret>
++}
+
++model Secret {
++  @doc("The resource id for the secret containing credentials for git.")
++  secret?: string;
++}
+```
+In future if we want to support other types of authentication(e.g ssh), `gitCredentials` would include another level of properties to specify the kind of authentication used.
+
+```
+model TerraformConfig{
+  @doc(Specifies git credentials information for terraform module repository.)  
+  gitCredentials?: Record<Record<Secret>>
+}
```

Comment:
This doesn't sound right. We anticipate that other kinds of auth will be needed in the future, so we can plan now to avoid a breaking change to the API.